accepted/future-releases/abstract-external-fields/feature-specification.md

```diff
@@ -0,0 +1,240 @@
+may arise, and a non-local variable declaration is a concise and
+non-redundant way to specify the desired signatures.
+
+This need came up in connection with `dart:ffi`, where users write Dart
+classes in order to specify a native memory layout. All instances of the
+interface will be backed by native (external) code. Currently, `dart:ffi`
+developers use a regular instance variable declaration, like `int foo;`,
+and add metadata to this declaration in order to specify the corresponding
+foreign language representation.
+
+What is actually needed is an external getter and possible setter, but if
+`dart:ffi` commits to using that then there will be a large amount of
+client code that declares getter/setter pairs with a non-trivial amount of
+redundancy: The name is specified twice, the type is specified twice, and
+there might be a need for two copies of metadata specifying the native
+representation. So `dart:ffi` uses regular instance variable declarations
+today.
+
+Similarly, a class can of course declare an abstract getter/setter pair if
+needed, but this again causes the name and type to be specified twice.
+
+As a workaround, it is possible to use a concrete instance variable as a
+replacement for an abstract getter and possibly setter, by simply ignoring
+the implementation. The concrete variable declaration will add the
+implicitly induced getter and possible setter to the interface of the
+enclosing class (let us call it `C`). Any concrete subtypes (`class D
+implements C ..`) will be required to implement said accessors. So far, it
+just works.
+
+However, if `D` is a _subclass_ of `C` then `D` is not required to
+implement the accessors. Instead, `D` tacity inherits the implementation of
+the concrete variable. This may be a bug, because `C` is designed for having
+these accessors overridden to do specific things that the concrete variable
+will not do. Also, if `D` _does_ override the accessors with a getter and
+setter as intended, the storage reserved for the concrete instance variable
+will be a space leak (it is unused).
+
+Finally, if null-safety is enabled then a concrete instance variable
+declaration (say, `int foo;`) is a compile-time error if the type of the
+variable is non-nullable, unless the variable is initialized in the
+initializer list of each generative constructor of the class. This problem
+arises both in connection with `dart:ffi` and for a an instance variable
+which is used as a replacement for an abstract getter and possibly setter.
+
+So the existing workaround of using a concrete instance variable to emulate
+an abstract instance variable is inconvenient, error prone, and confusing
+for a reader of the code, and for a writer of a subclass, and similar
+problems exist when a regular variable is used as a replacement for an
+external getter and possibly setter.
+
+
+## Design Idea
+
+In response to the issues above, this document introduces abstract and
+external variables.
```

Comment:
Maybe: ... and external variable declarations.
Since they don't actually introduce a *variable*.

We're not being consistent about this, so maybe not important.
(Personally, I don't think there is such a thing as an "abstract getter". There is an abstract getter declaration which introduces a getter signature and no getter implementation. There is no "getter" which is abstract in that.)

Reply:
Done in new PR on branch spec_external_abstract_fields_jun20.